File: server/internal/service/account.go
package service

import (
	"context"

	"cashcompass-server/internal/model"
	"cashcompass-server/internal/repository"
)

type AccountService interface {
	List(ctx context.Context) ([]model.Account, error)
	GetByID(ctx context.Context, id int) (model.Account, error)
	Create(ctx context.Context, label, accountType string) (model.Account, error)
	UpdateLabel(ctx context.Context, id int, label string) error
	UpdateType(ctx context.Context, id int, accountType string) error
	ToggleArchived(ctx context.Context, id int) (model.Account, error)
	Delete(ctx context.Context, id int) error
}

type AccountServiceImpl struct {
	accounts     repository.AccountRepository
	transactions repository.TransactionRepository
}

func NewAccountService(accounts repository.AccountRepository, transactions repository.TransactionRepository) *AccountServiceImpl {
	return &AccountServiceImpl{accounts: accounts, transactions: transactions}
}

func (s *AccountServiceImpl) List(ctx context.Context) ([]model.Account, error) {
	return s.accounts.List(ctx)
}

func (s *AccountServiceImpl) GetByID(ctx context.Context, id int) (model.Account, error) {
	return s.accounts.GetByID(ctx, id)
}

func (s *AccountServiceImpl) Create(ctx context.Context, label, accountType string) (model.Account, error) {
	return s.accounts.Create(ctx, label, accountType)
}

func (s *AccountServiceImpl) UpdateLabel(ctx context.Context, id int, label string) error {
	if err := s.accounts.UpdateLabel(ctx, id, label); err != nil {
		return err
	}
	if s.transactions != nil {
		return s.transactions.SyncAccountLabel(ctx, id, label)
	}
	return nil
}

func (s *AccountServiceImpl) UpdateType(ctx context.Context, id int, accountType string) error {
	return s.accounts.UpdateType(ctx, id, accountType)
}

func (s *AccountServiceImpl) ToggleArchived(ctx context.Context, id int) (model.Account, error) {
	a, err := s.accounts.GetByID(ctx, id)
	if err != nil {
		return model.Account{}, err
	}
	if err := s.accounts.SetArchived(ctx, id, !a.IsArchived); err != nil {
		return model.Account{}, err
	}
	return s.accounts.GetByID(ctx, id)
}

func (s *AccountServiceImpl) Delete(ctx context.Context, id int) error {
	return s.accounts.Delete(ctx, id)
}

File: src/services/account.py
from __future__ import annotations
from src.models import Account
from src.repository.account import AccountRepository
from src.repository.transaction import TransactionRepository


class AccountService:
    def __init__(self, repo: AccountRepository, txn_repo: TransactionRepository):
        self.repo = repo
        self.txn_repo = txn_repo

    @staticmethod
    def _normalize_label(label: str) -> str:
        normalized = label.strip()
        if not normalized:
            raise ValueError("account label is required")
        return normalized

    def list(self) -> list[Account]:
        return self.repo.list()

    def get_by_id(self, id: int) -> Account:
        return self.repo.get_by_id(id)

    def create(self, label: str, account_type: str) -> Account:
        return self.repo.create(self._normalize_label(label), account_type)

    def update_label(self, id: int, label: str):
        label = self._normalize_label(label)
        self.repo.update_label(id, label)
        self.txn_repo.sync_account_label(id, label)

    def update_type(self, id: int, account_type: str):
        self.repo.update_type(id, account_type)

    def toggle_archived(self, id: int) -> Account:
        acct = self.repo.get_by_id(id)
        self.repo.set_archived(id, not acct.is_archived)
        return self.repo.get_by_id(id)

    def delete(self, id: int):
        self.repo.delete(id)
